rs/consensus/vetkd/src/lib.rs

```diff
@@ -426,3 +428,693 @@ fn reject_if_invalid(
+        };
+        test_payload_builder(Some(config), contexts, shares, |builder| {
+            let payload = builder.build_payload(HEIGHT, MAX_SIZE, &[], &context);
+            assert!(payload.is_empty());
+
+            // Non-empty payload validation should be fail if we don't have the state
+            let payload = as_bytes(make_vetkd_agreements([0, 1, 2]));
+            let validation = builder.validate_payload(
+                HEIGHT,
+                &ProposalContext {
+                    proposer: node_test_id(0),
+                    validation_context: &context,
+                },
+                &payload,
+                &[],
+            );
+            assert_matches!(
+                validation.unwrap_err(),
+                ValidationError::ValidationFailed(
+                    PayloadValidationFailure::VetKdPayloadValidationFailed(
+                        VetKdPayloadValidationFailure::StateUnavailable(_)
+                    )
+                )
+            );
+        })
+    }
+
+    #[test]
+    fn test_build_empty_payload_if_pool_is_empty() {
+        let config = make_chain_key_config();
+        let contexts = make_contexts(&config);
+        test_payload_builder(Some(config), contexts, vec![], |builder| {
+            let payload = builder.build_payload(HEIGHT, MAX_SIZE, &[], &VALIDATION_CONTEXT);
+            assert!(payload.is_empty());
+
+            // Empty payloads should always be valid
+            let validation = builder.validate_payload(
+                HEIGHT,
+                &ProposalContext {
+                    proposer: node_test_id(0),
+                    validation_context: &VALIDATION_CONTEXT,
+                },
+                &payload,
+                &[],
+            );
+            assert!(validation.is_ok());
+        })
+    }
+
+    #[test]
+    #[should_panic(expected = "not yet implemented")]
+    fn test_build_empty_payload_max_size_zero() {
+        let config = make_chain_key_config();
+        let contexts = make_contexts(&config);
+        let shares = make_shares(&contexts);
+        test_payload_builder(Some(config), contexts, shares, |builder| {
+            let payload =
+                builder.build_payload(HEIGHT, NumBytes::from(0), &[], &VALIDATION_CONTEXT);
+            assert!(payload.is_empty());
+        })
+    }
+
+    #[test]
+    #[should_panic(expected = "not yet implemented")]
+    fn test_build_payload_respects_max_size() {
+        let config = make_chain_key_config();
+        let contexts = make_contexts(&config);
+        let shares = make_shares(&contexts);
+        test_payload_builder(Some(config), contexts, shares, |builder| {
+            // Use a small maximum size
+            let payload =
+                builder.build_payload(HEIGHT, NumBytes::from(50), &[], &VALIDATION_CONTEXT);
+            let payload_deserialized = bytes_to_vetkd_payload(&payload).unwrap();
+            assert_eq!(payload_deserialized.vetkd_agreements.len(), 1);
```

Comment:
I think it makes sense to assert here, that the serialized payload is smaller than 50 bytes.